src/Microsoft.AspNetCore.OData/Routing/ODataActionSelector.cs

```diff
@@ -126,31 +162,90 @@ public ActionDescriptor SelectBestCandidate(RouteContext context, IReadOnlyList<
+                    // and end up returning DoSomething(int key) instead of DoSomething()
+                    bestCandidate = matchedCandidates.FirstOrDefault(candidate => candidate.FilteredParameters.Count() == availableKeysCount);
+                    if (bestCandidate != null)
+                    {
+                        return bestCandidate.ActionDescriptor;
+                    }
+                }
 
-                // Return either the best matched candidate or the first
-                // candidate if none matched.
-                return (matchedCandidates.Any())
-                    ? candidates.Where(c => c.Id == matchedCandidates.FirstOrDefault().Id).FirstOrDefault()
-                    : candidates.FirstOrDefault();
+                return matchedCandidates.Select(c => c.ActionDescriptor).FirstOrDefault();
             }
 
             return _innerSelector.SelectBestCandidate(context, candidates);
         }
 
-        private bool TryMatch(IList<ParameterDescriptor> parameters, IList<string> availableKeys, ODataOptionalParameter optionalWrapper)
+        private bool TryMatch(
+            RouteContext context,
+            IList<ParameterDescriptor> parameters,
+            IList<string> availableKeys,
+            ODataOptionalParameter optionalWrapper,
+            int totalParameterCount,
+            int availableKeysCount)
         {
+            // navigationProperty is optional in some cases, therefore an action
+            // should not be rejected simply because it does not declare a navigationProperty parameter
+            if (availableKeys.Contains(ODataRouteConstants.NavigationProperty.ToLowerInvariant()))
+            {
+                availableKeysCount -= 1;
+            }
+
+            // reject action if it doesn't declare a parameter for each segment key
+            // e.g. Get() will be rejected for route /Persons/1
+            if (totalParameterCount < availableKeysCount)
+            {
+                return false;
+            }
+
+            bool matchedBody = false;
+            var conventionsStore = context.HttpContext.ODataFeature().RoutingConventionsStore;
             // use the parameter name to match.
-            foreach(var p in parameters)
+            foreach (var p in parameters)
             {
                 string parameterName = p.Name.ToLowerInvariant();
                 if (availableKeys.Contains(parameterName))
                 {
                     continue;
                 }
 
+                if (conventionsStore != null)
+                {
+                    // the convention store can contain the parameter as key
+                    // with a nested property (e.g. customer.Name) 
+                    if (conventionsStore.Keys.Any(k => k.Contains(p.Name)))
+                    {
+                        continue;
+                    }
+                }
+
+                if (context.HttpContext.Request.Query.ContainsKey(p.Name))
```

Comment:
@habbes Do we care about case here?

Reply:
From the documentation, I can't tell for certain if the comparison is case-sensitive https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.querycollection.containskey?view=aspnetcore-3.1